### PR TITLE
[CRYPTO-88] Implement Idempotency-Key header in Fireblocks

### PR DESCRIPTION
**Pull request by gracetrust**
Gem update [**Crypto-88**](https://primetrustspc.atlassian.net/browse/CRYPTO-88?atlOrigin=eyJpIjoiYmZkMmMzNWQ4OTRmNDAwZDkyMGNkY2U1ODM2MGQzMjYiLCJwIjoiaiJ9)
[Here](https://github.com/fundamerica/prime_trust_api/pull/2771) is API portion.

### CODE REVIEW

File: lib/fireblocks/api/request.rb [hunk 1]
```diff
@@ -55,7 +55,8 @@ def request_headers(body)
       {
         'X-API-Key' => Fireblocks.configuration.api_key,
         'Authorization' => "Bearer #{token(body)}",
-        'Content-Type' => 'application/json'
+        'Content-Type' => 'application/json',
+        'Idempotency-Key' => SecureRandom.uuid
```
Comment (mculp) [inline lib/fireblocks/api/request.rb:59]: This is almost exactly what we need.. but we need a way of knowing/remembering this key so we can send it on subsequent requests.

Comment (gracetrust) [reply lib/fireblocks/api/request.rb:59]: Are we planning to store the keys using Redis?

Comment (mculp) [reply lib/fireblocks/api/request.rb:59]: No, maybe we should pass the key in. Maybe it should be the `onchain_transaction_id`

Comment (mculp) [reply lib/fireblocks/api/request.rb:59]: In this file, we need to:

1) update `self.post` to accept `headers: {}`
2) update `post` to accept `headers: {}`
3) inside of `post` (not `self.post`), merge `request_headers` w/ `headers`

File: fireblocks.gemspec [hunk 2]
```diff
@@ -5,7 +5,7 @@ $LOAD_PATH.unshift(lib) unless $LOAD_PATH.include?(lib)
 require 'fireblocks/version'
 
 Gem::Specification.new do |spec|
-  spec.required_ruby_version = '>= 2.3.0'
+  spec.required_ruby_version = '>= 2.4'
 
   spec.name          = 'fireblocks'
   spec.version       = Fireblocks::VERSION
```
Comment (mculp) [inline fireblocks.gemspec:11]: Don't forget to update `Fireblocks::VERSION` in `lib/fireblocks/version.rb`

File: lib/fireblocks/api/request.rb [hunk 3]
```diff
@@ -41,21 +41,24 @@ def put(body)
...
     private
 
-    def request_headers(body)
+    def request_headers(body, headers = {})
       {
         'X-API-Key' => Fireblocks.configuration.api_key,
         'Authorization' => "Bearer #{token(body)}",
-        'Content-Type' => 'application/json'
+        'Content-Type' => 'application/json',
+        'Idempotency-Key' => headers['Idempotency-Key']
```
Comment (snogrammer) [inline lib/fireblocks/api/request.rb:61]: this is going to be sending an empty header value if this is not defined. Is that going to cause issues?

File: lib/fireblocks/api/transactions.rb [hunk 4]
```diff
@@ -9,21 +9,22 @@ class << self
...
-          autoStaking
           networkStaking
-          cpuStaking
+          note
+          source
+          headers
         ].freeze
 
-        def create(options)
+        def create(options, headers)
```
Comment (snogrammer) [inline lib/fireblocks/api/transactions.rb:25]:
```suggestion
        def create(options, headers: {})
```

File: lib/fireblocks/api/request.rb [hunk 5]
```diff
@@ -41,21 +41,24 @@ def put(body)
...
+      @idempotency_key = idempotency_key
       req.body = body.to_json
 
       valid_response!(send_request(req), request: req)
     end
 
     private
 
-    def request_headers(body)
+    def request_headers(body, headers = {})
```
Comment (snogrammer) [inline lib/fireblocks/api/request.rb:56]: Keyword args would be better here 

```suggestion
    def request_headers(body, headers: {})
```

File: lib/fireblocks/api/request.rb [hunk 6]
```diff
@@ -41,21 +41,24 @@ def put(body)
       valid_response!(send_request(req), request: req)
     end
 
-    def post(body)
+    def post(body, headers)
```
Comment (snogrammer) [inline lib/fireblocks/api/request.rb:44]: keyword arg:

```suggestion
    def post(body, headers: {})
```

File: lib/fireblocks/api/request.rb [hunk 7]
```diff
@@ -41,21 +41,24 @@ def put(body)
...
       req = Net::HTTP::Post.new(uri)
-      request_headers(body).each { |rk, rv| req[rk] = rv }
+      request_headers(body, headers).each { |rk, rv| req[rk] = rv }
+      req['Idempotency-Key'] = idempotency_key if idempotency_key
+      @idempotency_key = idempotency_key
```
Comment (snogrammer) [inline lib/fireblocks/api/request.rb:48]: This is being set both here and L61. 

What is the purpose of setting a `idempotency_key` attr_accessor?

File: lib/fireblocks/api/request.rb (hunk 5 above)
Comment (gracetrust) [reply lib/fireblocks/api/request.rb:54]: @snogrammer I reverted back to default args bc the `idempotency-key` wasn't being passed through headers with the keyword args

**Review by snogrammer — APPROVED** [review]

**Pull request merged**